servicetalk-dns-discovery-netty/src/main/java/io/servicetalk/dns/discovery/netty/DefaultDnsClient.java

```diff
@@ -0,0 +1,715 @@
+        return asyncCloseable.closeAsyncGracefully();
+    }
+
+    private void closeAsync0() {
+        assertInEventloop();
+
+        if (closed) {
+            return;
+        }
+        closed = true;
+        resolver.close();
+        ttlCache.clear();
+    }
+
+    private void assertInEventloop() {
+        assert nettyIoExecutor.isCurrentThreadEventLoop() : "Must be called from the associated eventloop.";
+    }
+
+    private final class SrvRecordPublisher extends AbstractDnsPublisher<HostAndPort> {
+        private final String serviceName;
+
+        private SrvRecordPublisher(String serviceName) {
+            this.serviceName = requireNonNull(serviceName);
+        }
+
+        @Override
+        public String toString() {
+            return serviceName;
+        }
+
+        @Override
+        protected AbstractDnsSubscription newSubscription(
+                final Subscriber<? super Iterable<ServiceDiscovererEvent<HostAndPort>>> subscriber) {
+            return new AbstractDnsSubscription(true, subscriber) {
+                @Override
+                protected Future<DnsAnswer<HostAndPort>> doDnsQuery() {
+                    Promise<DnsAnswer<HostAndPort>> promise = ImmediateEventExecutor.INSTANCE.newPromise();
+                    resolver.resolveAll(new DefaultDnsQuestion(serviceName, SRV))
+                            .addListener((Future<? super List<DnsRecord>> completedFuture) -> {
+                                Throwable cause = completedFuture.cause();
+                                if (cause != null) {
+                                    promise.setFailure(cause);
+                                } else {
+                                    final DnsAnswer<HostAndPort> dnsAnswer;
+                                    long minTTLSeconds = Long.MAX_VALUE;
+                                    try {
+                                        @SuppressWarnings("unchecked")
+                                        final List<DnsRecord> dnsRecords = (List<DnsRecord>) completedFuture.getNow();
+                                        final List<HostAndPort> hostAndPorts = new ArrayList<>(dnsRecords.size());
+                                        for (DnsRecord dnsRecord : dnsRecords) {
+                                            if (!SRV.equals(dnsRecord.type()) || !(dnsRecord instanceof DnsRawRecord)) {
+                                                throw new IllegalArgumentException(
+                                                        "Unsupported DNS record type for SRV query: " + dnsRecord);
+                                            }
+                                            if (dnsRecord.timeToLive() < minTTLSeconds) {
+                                                minTTLSeconds = dnsRecord.timeToLive();
+                                            }
+                                            ByteBuf content = ((DnsRawRecord) dnsRecord).content();
```

Comment:
@Scottmitch I think you also need to call `ReferenceCounted.release(dnsRecord)`here.

Reply:
Fixed in https://github.com/apple/servicetalk/pull/1044